Setup basic configuration

slfuchs opened the pull request:
Setup the basic configuration of PartExa making use of provided deal.II functionality. This includes the configuration of clang-format, cmake, and the testing framework.

slfuchs committed:
Setup basic cmake and testing configuration

Setup a basic cmake configuration for the library PartExa
(including deal.II) and a testing configuration.

Introduce a PartExa header that is printed during configuration and
printed to the output stream by a function. Add test for this function.

slfuchs committed:
Add clang-format configuration based on deal.II

slfuchs commented on .clang-format:
```diff
@@ -0,0 +1,187 @@
+BreakConstructorInitializers: BeforeComma
+BreakStringLiterals: false
+
+ColumnLimit: 120
```
@ppraegla, @nfehn, @peterrum: What would be a good choice for the column limit in your opinion?

Most programs traditionally use 80 characters. However, nowadays there is a trend towards 100 characters or even more. I personally would prefer 120 characters, as I find code with excessive line breaks hard to read.

peterrum replied:
Don't ask me. I am strong advocate of 80 ;)

nfehn replied:
In ExaDG we used column-limit 100. For me, this is a good compromise, as it gives motivation to write nice code or refactor the code as compared to 120 characters (but I fully agree that 80 characters is too narrow). But we can also start with 120.

Overall, this is not so important for me regarding code formatting.

ppraegla replied:
I would also go for 100. 120 seems a lot.

slfuchs replied:
Ok, let's go to 100. That seems to be a good compromise.

slfuchs commented on CMakeLists.txt:
```diff
@@ -0,0 +1,79 @@
+
+INCLUDE_DIRECTORIES(include)
+
+#ADD_LIBRARY(partexa
+#  todo: add source files
+#  )
+#DEAL_II_SETUP_TARGET(partexa)
```
This section is currently commented out, as no source files are implemented yet. However, later on it can easily be added to cmake.

peterrum replied:
I would suggest to add a `dummy.cc` file. See deal.II or hyper.deal. After that I would add GitHub CI

slfuchs commented on tests/base/print_partexa_header.output:
```diff
@@ -0,0 +1,12 @@
+
+
+________________________________________________________________________________
+                       ____           _   _____
+                      |  _ \__ _ _ __| |_| ____|_  ____ _
+                      | |_)| _` | '__| __|  _| \ \/ / _` |
+                      |  __/(_| | |  | |_| |___ >  < (_| |
+                      |_|  \__,_|_|   \__|_____/_/\_\__,_|
+
+                      A Particle Library for the Exa-Scale
+________________________________________________________________________________
```
This is the fancy `PartExa` header that is printed during configuration or when calling the method defined in `include/base/print_infos.h`.

Edit: It looks way better in a terminal. The line spacing in GitHub is not optimal for ASCII Art.

nfehn replied:
Looks good! Maybe I would break the top line between the upper right part of the "x" and the final "a" (just as you did this for the bottom line.

nfehn replied:
Maybe the diagonal line of the "t" should start one character earlier (and the bottom line would then also be one character longer).

slfuchs commented:
@ppraegla, @nfehn, @peterrum: This pull request adds a very basic configuration following the [documentation](https://www.dealii.org/developer/index.html) of deal.II. This can be seen as a starting point for further development. In a next step, I will work on a brief documentation in the wiki herein.

peterrum commented on tests/base/CMakeLists.txt:
```diff
@@ -0,0 +1 @@
+DEAL_II_PICKUP_TESTS()
```
I would test explicitly the `partpxa` library here similarly as in https://github.com/hyperdeal/hyperdeal/blob/9bfd4b2189fe27896a782c641a30bda219437ff5/tests/base/CMakeLists.txt#L3

nfehn commented on cmake/macros/print_partexa_header.cmake:
```diff
@@ -0,0 +1,34 @@
+MESSAGE("                      | |_)| _` | '__| __|  _| \\ \\/ / _` |")
+MESSAGE("                      |  __/(_| | |  | |_| |___ >  < (_| |")
+MESSAGE("                      |_|  \\__,_|_|   \\__|_____/_/\\_\\__,_|")
+MESSAGE("")
```
I would probably fill up with spaces because this looks a bit strange

nfehn commented on include/base/print_infos.h:
```diff
@@ -0,0 +1,51 @@
+    << "                      | |_)| _` | '__| __|  _| \\ \\/ / _` |" << std::endl
+    << "                      |  __/(_| | |  | |_| |___ >  < (_| |" << std::endl
+    << "                      |_|  \\__,_|_|   \\__|_____/_/\\_\\__,_|" << std::endl
+    << "" << std::endl
```
the same here and for the lines above

slfuchs commented:
I included all suggestions made so far and pushed the changes. Currently I am waiting for the outcome of dealii/dealii/pull/13033 as it would allow us to remove the copy of the `tests.h` file.

peterrum commented:
> I included all suggestions made so far and pushed the changes. Currently I am waiting for the outcome of dealii/dealii/pull/13033 as it would allow us to remove the copy of the tests.h file.

I wouldn't wait.

slfuchs commented:
@peterrum I added now the reduced version of the `tests.h` file. In addition to what you prepared I included the header `<fstream>` and added the definition `void cat_file(const char *filename)`.

slfuchs commented on tests/base/print_partexa_header.cc:
```diff
@@ -0,0 +1,36 @@
+
+#include "../tests.h"
+
+int
+main()
+{
+  dealii::initlog();
+
+  dealii::ConditionalOStream pcout(dealii::deallog.get_file_stream(), true);
+
+  PartExa::print_partexa_header(pcout);
+
+  return 0;
+}
```
@peterrum: Do you prefer to make use of `using namespace dealii;` and `using namespace PartExa` here?

peterrum replied:
I don't really care. Personally, I prefer to be explicit regarding namespaces (i.e. not use `using namespace dealii;`) - this is what we try to do in hyper.deal (except in the test folder^^). The PartExa project is still at the beginning so banning the usage of `using namespace dealii;` is still possible without too much headache. `using namespace PartExa` is reasonable; these are the tests of PartExa^^

nfehn replied:
I would aim to not use `using namespace dealii;`. For example, I would try to avoid this when writing ExaDG once more. I think the code is overall easier to understand if one immediately sees where a certain functionality originates from.

slfuchs replied:
Thanks, its good to hear that both of you confirm that. So let's do that, and don't use `using namespace dealii;` in PartExa. I add that to the contribution guidelines as soon as I found the time to set up a wiki.

peterrum commented on tests/tests.h:
```diff
@@ -0,0 +1,159 @@
+
+DEAL_II_NAMESPACE_OPEN
+
+// given the name of a file, copy it to deallog
+// and then delete it
+void
+cat_file(const char *filename)
```
If you think this function is useful you can mention it in https://github.com/dealii/dealii/pull/13033.

slfuchs replied:
This function is used in `struct MPILogInitAll` so must be defined otherwise one gets a build error.

slfuchs replied:
I write a short comment about that in dealii/dealii#13033.

slfuchs commented:
Thanks for the feedback to all. I included everything as discussed above and merge this pull request now.
